WalletWasabi/Crypto/ZeroKnowledge/Statement.cs

```diff
@@ -0,0 +1,31 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge
+{
+	public class Statement
+	{
+		public Statement(GroupElement publicPoint, IEnumerable<GroupElement> generators)
```

Comment:
In order to ensure different proofs are differentiated even when sharing the same generators and referring to the same public input(s), `Statement` should also have a `string` (as UTF8 i guess) or `byte[]` tag/identifier, which could be either the proof's name in the protocol or the statement copied from the LaTeX, or both.

If `Statement` is intended to model the entire statement then the elemntary subproofs can simply share a single tag between them identifying the conjunction.

If `Statement` is intended to model elementary statements so that compound proofs will form a tree of `Statement` objects, then recording a trace of the recursive descent would ensure that the higher structure is also committed.

Reply:
Good idea. Will see what to do with that when the issue comes up.